diff -Nrup a/read.c b/read.c
--- a/read.c	2015-12-17 00:18:34.549075980 -0500
+++ b/read.c	2015-12-17 00:25:45.531805445 -0500
@@ -459,7 +459,10 @@ eval_buffer (char *buffer)
   ebuf.buffer = ebuf.bufnext = ebuf.bufstart = buffer;
   ebuf.fp = NULL;
 
-  ebuf.floc = *reading_file;
+  if (reading_file)
+    ebuf.floc = *reading_file;
+  else
+    ebuf.floc.filenm = NULL;
 
   curfile = reading_file;
   reading_file = &ebuf.floc;
