diff -Nrup a/function.c b/function.c
--- a/function.c	2006-04-01 01:36:40.000000000 -0500
+++ b/function.c	2015-12-17 10:29:59.247336412 -0500
@@ -1663,7 +1663,15 @@ func_shell (char *o, char **argv, const 
   if (pid < 0)
     perror_with_name (error_prefix, "fork");
   else if (pid == 0)
-    child_execute_job (0, pipedes[1], command_argv, envp);
+    {
+#ifdef SET_STACK_SIZE
+      /* Reset limits, if necessary.  */
+      if (stack_limit.rlim_cur)
+	setrlimit (RLIMIT_STACK, &stack_limit);
+#endif
+
+      child_execute_job (0, pipedes[1], command_argv, envp);
+    }
   else
 
 # endif
